mpi4jax/_src/tokenizer.py

```diff
@@ -0,0 +1,60 @@
+import jax
+from jax._src.util import safe_map
+from jax import linear_util as lu
+token_override_registry = {}
+
+def _override_tokens(jaxpr, consts, token, *args):
+    if token is None: # Create a new token if one is not passed.
+        token = jax.lax.create_token()
+    def read(v):
+        if type(v) is jax.core.Literal:
+            return v.val
+        else:
+            return env[v]
+
+    def write(v, val):
+        env[v] = val
+    
+    env = {}
+    write(jax.core.unitvar, jax.core.unit)
+    safe_map(write, jaxpr.constvars, consts)
+    safe_map(write, jaxpr.invars, args)
+    for eqn in jaxpr.eqns:
+        # Here, we override the original primatives with our own
+        # token forcing method.
+        if eqn.primitive in token_override_registry:
+            token_override = token_override_registry[eqn.primitive]
+            ans = token_override(
+                safe_map(read, eqn.invars), new_token=token, **eqn.params)
+            # We pass along the new token returned by the above binding
+            token = ans[-1]
+            safe_map(write, eqn.outvars, ans)
+        else:
+            # Here, we are just reapplying the original operation if 
+            # not part of our communication protocol. 
+            # This code is mostly taken form jax.core.eval_jaxpr
+            subfuns, bind_params = eqn.primitive.get_bind_params(eqn.params)
+            subfuns = safe_map(lambda x: lu.wrap_init(_auto_tokenize(x.call_wrapped)), subfuns)
```

Comment:
I think that in order to properly support jitted function here you should call `auto_tokenize(x.call_wrapped, token=token)` and somehow get the resulting token out of this function?

Reply:
to make this work, you probably should return the current `token` from the function transformed by `_override_tokens` so that it can be properly be weaved through subsequent subfunction calls.

Then it can be later stripped at the level of the `auto_tokenize` wrapper.

Reply:
Yeah that was originally my plan, but for some reason I kept getting the error

```
TypeError: No constant handler for type: <class 'jax.core.Token'>
```

Which I was unable to debug.

Reply:
Anything for you Matt.

Reply:
https://github.com/google/jax/issues/9438